src/Features/TerrainHelper.cpp

```diff
@@ -0,0 +1,134 @@
+#include "TerrainHelper.h"
+
+#include "ShaderCache.h"
+#include "State.h"
+
+void TerrainHelper::DataLoaded()
+{
+	// Get the default landscape texture set for terrain helper
+	const auto defaultLandTextureSet = RE::TESForm::LookupByEditorID<RE::BGSTextureSet>("LandscapeDefault");
+	if (defaultLandTextureSet != nullptr) {
+		spdlog::info("LandscapeDefault EDID texture set found");
+		defaultLandTexture = defaultLandTextureSet;
+	} else {
+		spdlog::info("LandscapeDefault EDID texture set not found, using default");
+		const auto bgsDefaultLandTex = *REL::Relocation<RE::TESLandTexture**>(RELOCATION_ID(514783, 400936));
+		defaultLandTexture = bgsDefaultLandTex->textureSet;
+	}
+}
+
+bool TerrainHelper::TESObjectLAND_SetupMaterial(RE::TESObjectLAND* land)
+{
+	if (land->loadedData == nullptr || land->loadedData->mesh[0] == nullptr) {
+		// this is not terrain or vanilla material failed
+		return false;
+	}
+
+	for (uint32_t quadI = 0; quadI < 4; ++quadI) {
+		// Get hash key of vanilla material
+		uint32_t hashKey = 0;
+
+		const auto& children = land->loadedData->mesh[quadI]->GetChildren();
```

Comment:
Nested ctds here. Please check nullptrs for each pointer before use.

Reply:
Updated